### DISCUSSION TITLE
How to do cumulative sum on an array?

### DISCUSSION THREAD

**Discussion by agent3133** [atoti/atoti#559] [Q&A] [answered]
Hi,

When I'm trying to create a cumulative sum on an array, it seems that values are not calculated correctly. 

![image](https://user-images.githubusercontent.com/24815479/163050726-197c8905-f850-4a11-b7e4-3655f8901d51.png)

Here is the sample of the code:

```python
import pandas as pd
import atoti as tt

session = tt.Session(port=5555)

df = pd.DataFrame(
   {
      'ids': ['a', 'b'],
      'vectors': [(1,2,3), (10,20,30)]
   }   
)

session = tt.Session('example_cum_sum')
example = session.read_pandas(df, table_name='Vector_Example')
cube = session.create_cube(example, 'example')
m, h, l = cube.measures, cube.hierarchies, cube.levels
m["ElementSum"] = m["vectors.SUM"][m["Date index"]]
m['CumulativeSum'] = tt.agg.sum(
    m['vectors.SUM'], scope=tt.scope.cumulative(l['Dates'])
)[m['Date index']]
```

Best Regards,
Alex

**Comment by bloussou** [comment, marked as answer]
Hello,
Which version of atoti are you using ?
Your example is missing `m['Date index']` and `l["Dates"]` can you clarify it?
I have correct result with the following code :
```python
import pandas as pd
import atoti as tt

session = tt.Session(port=5555)

df = pd.DataFrame(
   {
      'Dates': ["2021-03-01", "2021-04-01"],
      'vectors': [(1,2,3), (10,20,30)]
   }   
)

session = tt.Session()
example = session.read_pandas(df, table_name='Vector_Example')
cube = session.create_cube(example)
m, h, l = cube.measures, cube.hierarchies, cube.levels
m["ElementSum"] = m["vectors.SUM"][0]
m['CumulativeSum'] = tt.agg.sum(
    m['ElementSum'], scope=tt.scope.cumulative(l['Dates'])
)
m['CumulativeSumVector'] = tt.agg.sum(
    m["vectors.SUM"], scope=tt.scope.cumulative(l['Dates'])
)[0]
```
![image](https://user-images.githubusercontent.com/32448554/163134677-32678c4d-3751-4089-bd00-4e4d7655cf1d.png)

What if you define `m['CumulativeSum']` like this:
```python
m['CumulativeSum'] = tt.agg.sum(
    m['ElementSum'], scope=tt.scope.cumulative(l['Dates'])
)
```

**Reply by agent3133** [reply]
Hi [bloussou](https://github.com/bloussou), 

thanks for the analysis. 

Regarding the l['Dates'] and m['Date index']. I have created it as below, and it is important for the example, because I want to calculate cumulative sum along the array, not along the Ids column.

`cube.create_parameter_hierarchy_from_members('Dates', 
                                              members=[pd.to_datetime('20211231'), 
                                                       pd.to_datetime('20221231'), 
                                                       pd.to_datetime('20231231')],
                                              index_measure_name="Date index"
                                             )`

I'm using Atoti 0.6.5

Best regards

**Reply by agent3133** [reply]
A collegue of mine have find my mistake.
It should be 

```
m['CumulativeSum'] = tt.agg.sum(
    m['ElementSum'], scope=tt.scope.cumulative(l['Dates'])
)
```

**Reply by rcolle** [reply]
Alternatively, if your array elements are ordered by dates, you could use a prefix sum (https://docs.atoti.io/latest/lib/atoti/atoti.array.prefix_sum.html) followed by the index operation. 
Something along the lines of 
```python
m['CumulativeSum'] = tt.array.prefix_sum(m['vectors.SUM'])[m['Date index']]
```

**Reply by agent3133** [reply]
> 

Oh cool it's looks indeed cleaner
